wordpress-mobile/WordPress-iOS
# Use persistent storage for @-mentions suggestions

This changes the storage of @-mentions from an in-memory cache to a persistent Core Data store. This will allow us to provide an offline experience for @-mentions in the Gutenberg block editor.

Update: Other changes include updating `SuggestionService` to use a callback-based rather than a notification-based mechanism. 

Addresses https://github.com/wordpress-mobile/gutenberg-mobile/issues/2577

@-mentions are used in three places in the app: Reader comments, Notification comments/replies, and the Gutenberg block editor.

Known issues:
- Reader: @-mentions not working on Atomic sites: https://github.com/wordpress-mobile/WordPress-iOS/issues/14844 (**existing bug**, not related to this PR)
- ~~@-mentions data is shared between users on the same device: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2600~~ no longer the case on iOS after https://github.com/wordpress-mobile/WordPress-iOS/commit/2c343d7e4c10bde8b51947d6af146fae7812c7f4
- Supporting @-mentions on Jetpack connected self-hosted sites: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2601
- Even if multiple sites return (via `rest/v1.1/users/suggest`) largely the same suggestions list, the app stores each list separately in Core Data. In one test where the user visited 5 sites each with ~1500 results, Core Data contained ~6500 `AtMentionSuggestion` records. However, these records are deleted when the user logs out.
- If the suggestion list changes on the server (e.g. a user is added, deleted or edited), the app will not see the change if it has already fetched a non-empty suggestion list for that site, or until the user logs back in again

## How to test

The goal in testing here is to verify that the app behaves as it did before this change.

#### Test checklist template

```
- [ ] Reader comments
- [ ] Comment on post (via notification)
- [ ] Reply to comment (via notification)
- [ ] Gutenberg editor mentions
```

**What you'll need**: A WordPress.com site (may be your own site) that you can comment on and that has one or more other WordPress.com users (in order to test @-mentions). Make sure some of the users have a profile photo. The user that you login with will also need:
- a notification about a post
- a notification about a comment

### Reader comments

<details>
<summary>Test steps</summary>
<ol>
<li>In the Reader tab, locate the site</li>
<li>Navigate to a post, tap the comment button</li>
<li>Tap the &quot;Reply to post...&quot; text box to bring up the comment UI</li>
<li>Type <code>@</code> and verify that the list of suggestions is shown</li>
<li>Make sure that the avatar images load for users who have a profile photo</li>
<li>Select one of the suggestions to add it to your comment and hit send</li>
<li>Verify that the mentioned user receives a notification</li>
</ol>
</details>

### Comment on post (via notification)

<details>
<summary>Test steps</summary>
<ol>
<li>In the Notifications tab, locate a notification about a post</li>
<li>Open it, then tap the comment button, then the &quot;Reply to post...&quot; text box to bring up the comment UI</li>
<li>Type <code>@</code> and verify that the list of suggestions is shown</li>
<li>Make sure that the avatar images load for users who have a profile photo</li>
<li>Select one of the suggestions to add it to your comment and hit send</li>
<li>Verify that the mentioned user receives a notification</li>
</ol>
</details>

### Reply to comment (via notification)

<details>
<summary>Test steps</summary>
<ol>
<li>In the Notifications tab, locate a notification about a comment</li>
<li>Open it, then tap the &quot;Write a reply...&quot; text box to bring up the comment UI</li>
<li>Type <code>@</code> and verify that the list of suggestions is shown</li>
<li>Make sure that the avatar images load for users who have a profile photo</li>
<li>Select one of the suggestions to add it to your reply and hit send</li>
<li>Verify that mentioned user receives a notification</li>
</ol>
</details>

### Gutenberg editor mentions

<details>
<summary>Test steps</summary>
<ol>
<li>Open the block editor</li>
<li>In the &quot;Start writing...&quot; section, type <code>@</code> and verify that the list of suggestions is shown</li>
<li>Make sure that the avatar images load for users who have a profile photo</li>
<li>Select one of the suggestions and publish the post</li>
<li>Verify that mentioned user receives a notification</li>
</ol>
</details>

In addition to the above tests, I also did extra testing (which is not necessary for PR review):
- filtering names by typing after the @ character and ensuring the list is filtered
- tested on sites that contain hundreds (or thousands) of users
- fresh installs for each of the above tests to make sure nothing is carried over from one test to the next
- Switch the device offline after doing the above tests, and ensure the offline (cached) results are shown
- Tested edge case where an empty list of suggestions is retrieved (e.g. when a user has the role Contributor on a site, that user is not able to retrieve a list of suggestions)

---

PR submission checklist:
